internal/transformer/transformresult.go

```diff
@@ -0,0 +1,377 @@
+// Copyright (C) 2018 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package transformer
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math"
+	"strconv"
+
+	"github.com/edgexfoundry/device-sdk-go/internal/common"
+	ds_models "github.com/edgexfoundry/device-sdk-go/pkg/models"
+	"github.com/edgexfoundry/edgex-go/pkg/models"
+)
+
+const (
+	defaultBase   string = "0"
+	defaultScale  string = "1.0"
+	defaultOffset string = "0.0"
+)
+
+func TransformReadResult(cv *ds_models.CommandValue, pv models.PropertyValue) error {
+	if cv.Type == ds_models.String || cv.Type == ds_models.Bool {
+		return nil // do nothing for String and Bool
+	}
+
+	value, err := commandValueForTransform(cv)
+	newValue := value
+
+	if pv.Base != "" && pv.Base != defaultBase {
+		newValue, err = transformReadBase(newValue, pv.Base)
+		if err != nil {
+			return err
+		}
+	}
+
+	if pv.Scale != "" && pv.Scale != defaultScale {
+		newValue, err = transformReadScale(newValue, pv.Scale)
+		if err != nil {
+			return err
+		}
+	}
+
+	if pv.Offset != "" && pv.Offset != defaultOffset {
+		newValue, err = transformReadOffset(newValue, pv.Offset)
+	}
+
+	if value != newValue {
+		err = replaceNewCommandValue(cv, newValue)
+	}
+	return err
+}
+
+func transformReadBase(value interface{}, base string) (interface{}, error) {
+	b, err := strconv.ParseFloat(base, 64)
+	if err != nil {
+		common.LoggingClient.Error(fmt.Sprintf("the base %s of PropertyValue cannot be parsed to float64: %v", base, err))
+		return value, err
+	} else if b == 0 {
+		return value, nil // do nothing if Base = 0
+	}
+
+	var valueFloat64 float64
+	switch v := value.(type) {
+	case uint8:
+		valueFloat64 = float64(v)
+	case uint16:
+		valueFloat64 = float64(v)
+	case uint32:
+		valueFloat64 = float64(v)
+	case uint64:
+		valueFloat64 = float64(v)
+	case int8:
+		valueFloat64 = float64(v)
+	case int16:
+		valueFloat64 = float64(v)
+	case int32:
+		valueFloat64 = float64(v)
+	case int64:
+		valueFloat64 = float64(v)
+	case float32:
+		valueFloat64 = float64(v)
+	case float64:
+		valueFloat64 = v
+	}
+
+	valueFloat64 = math.Pow(b, valueFloat64)
+
+	switch value.(type) {
+	case uint8:
+		value = uint8(valueFloat64)
+	case uint16:
+		value = uint16(valueFloat64)
+	case uint32:
+		value = uint32(valueFloat64)
+	case uint64:
+		value = uint64(valueFloat64)
+	case int8:
+		value = int8(valueFloat64)
+	case int16:
+		value = int16(valueFloat64)
+	case int32:
+		value = int32(valueFloat64)
+	case int64:
+		value = int64(valueFloat64)
+	case float32:
+		value = float32(valueFloat64)
+	case float64:
+		value = valueFloat64
+	}
+	return value, err
+}
+
+func transformReadScale(value interface{}, scale string) (interface{}, error) {
+	switch v := value.(type) {
+	case uint8:
+		s, err := strconv.ParseUint(scale, 10, 8)
+		if err != nil {
+			common.LoggingClient.Error(fmt.Sprintf("the scale %s of PropertyValue cannot be parsed to %T: %v", scale, v, err))
+			return value, err
+		}
+		ns := uint8(s)
+		value = v * ns
```

Comment:
What happens if the result is too large to fit into value? The device services requirements document states that similar to an assertion, the result string should be set to "Overflow failed for device resource: <name>". This applies to offsets too.

Reply:
opened #88 for this